src/puppetlabs/puppetdb/http/reports.clj

```diff
@@ -7,22 +7,40 @@
             [puppetlabs.puppetdb.middleware :refer [verify-accepts-json validate-query-params
                                                     wrap-with-paging-options]]))
 
+(defn restrict-app-to-hash
+  [app hash]
+  (comp app (partial http-q/restrict-query-to-hash hash)))
+
+(defn entity-responder
+  [version]
+  (fn respond-with-entity
+    [entity]
+    (fn response-producer
+      [{:keys [params globals paging-options] :as request}]
+      (let [{db :scf-read-db url-prefix :url-prefix} globals
+            {:strs [query]} params]
+        (produce-streaming-body entity version query paging-options db url-prefix)))))
+
 (defn query-app
   [version]
-  (app
-    [hash "events" &]
-    (comp (e/events-app version)
-          (partial http-q/restrict-query-to-report hash))
+  (let [respond-with-entity (entity-responder version)]
```

Comment:
would it be cleaner to parameterize entity-responder with [version entity] (dropping a level of nesting) and bind this with a partial instead?